### ISSUE TITLE
Data preparing

### ISSUE THREAD

**Issue by Nazarko12345** [learnml-today/object-detection-with-pytorch#21]
Question about training.
-Why so small batch_size?
-We use batch_size=3 and a lot of image transformation like resizing and add padding.Is it better to use batch_size = 1 without image transforming?
-Why we resize to min_size if we can just add padding?
-Also question overall about detection.If i have a task to detect peoples from some video should i choose some model ,dataset and train it?How can i be confident that it will work for this video?Also  more complicated dataset more our model  robust in real life?

**Comment by creotiv** [comment]
Sorry, missed this question.

- We use maximum batch that can handle our GPU. The optimal batch for this training is 8, but it's a relative thing that depends on data and in practice always tested, so there is no silver bullet.
Bigger batch gives better generalization, and if you use for example Batch Normalization needed for this layer to work properly(with batch=1 it will not work), smaller gives faster training but much spiky gradient
- It will be technically simpler but will give worse result (if the network will be able to train at all)
- We have min_/max_ size so images are not very different in size in another case we will get huge paddings which will make training much harder. It the same problem as with coordinates prediction, which we solve in SSD with cells to minimize distribution variance.

- When you need to train to detect people on videos in the best scenario you should get part of such video labeled for training. This need because:
   - different sizes of input data and scales of people will make the problem
   - different cameras/optics create different pixel distribution which makes the first few layers of feature extractor totally different(that's why you cant train straight with 3d rendered data). And that's another problem.
   - also, different occlusions that may appear on video and not exist in public datasets will also make problem.
To fix this you should use or dataset from data on which you will be run inference, or use heavy augmentation.
What augmentation u can use to increase test accuracy:
- photo augmentation (lightness, contrast, etc)
- shape augmentation (skewing, mirroring, cropping)
- optical augmentation - making you training data similar to optical params of the video
- noise augmentation - adding same noise/interference on video
- low level features distribution on video - train feature extractor in an unsupervised way so the first layers learn distribution from video and not from the dataset. Then freeze them, and then train on a dataset with all these augmentations above. But this depends on the level of noise on video so not always working. Another way is to include noise from the video (i mean summing with some percentage like 5-10% random frame from video with input image from the dataset) this showed to increase accuracy on test inference.
But you must remember that more augmentation you are adding the more time needed to model to converge. And also it's good to write somewhere for yourself what is working for which data and what not. For example, you add new augmentation which gave +0.005% accuracy but increase training twice in time. That totally not worth it, and that should be remembered,
